File: problemsPerDay/2024/10/3211. 生成不含相邻零的二进制字符串.py
from typing import List


class Solution:
    def validStrings(self, n: int) -> List[str]:
        res = []

        for i in range(1 << n):
            if i & (i >> 1):
                continue
            res.append(bin(i ^ ((1 << n) - 1))[2:].zfill(n))

        return res


if __name__ == "__main__":
    print(Solution().validStrings(2))
    print(Solution().validStrings(3))
    print(Solution().validStrings(4))

File: problemsPerDay/2024/10/3211. 生成不含相邻零的二进制字符串.ts
function validStrings(n: number): string[] {
    const res: string[] = [];

    const mark = (1 << n) - 1;

    for (let i = 0; i <= mark; i++) {
        const str = (i ^ mark).toString(2).padStart(n, "0");

        if ((i & (i >> 1)) === 0) {
            res.push(str);
        }
    }

    return res;
}

console.log(validStrings(2)); //
console.log(validStrings(3)); //
console.log(validStrings(4)); // ["0001", "0100", "1000", "1010", "0101"]
